### PR TITLE
Resource proofing with lagging elders

### PR DESCRIPTION
**Pull request by maqi**
This PR contains the work of making a joining_node resend ConnectionInfoReq and CandidateInfo to an elder in case of not having connection to it or hasn't started resource proofing with it. Which allows resource proofing to be carried out with lagging elders.
A test is also create to confirm that the update will work as expected.

### CODE REVIEW

File: src/states/elder/mod.rs [hunk 1]
```diff
@@ -1057,11 +1061,26 @@ impl Elder {
...
+                self.ignore_candidate_info_counter -= 1;
+                return;
+            }
+        }
         debug!(
             "{} Handling CandidateInfo from {}->{}.",
             self, old_pub_id, new_pub_id
         );
 
+        if self.peer_mgr.has_candidate_info() {
```
Comment (jeanphilippeD) [inline src/states/elder/mod.rs]: I do not think we want that here.
It seem the bug is either in `self.peer_mgr.handle_candidate_info`, or the match processing the result.

When we call this we already know this is a valid candidate we are processing since `self.chain.matching_candidate_target_interval` did not lead us to return early, so we probably should not disconnect it then.

File: src/states/proving_node.rs [hunk 2]
```diff
@@ -248,11 +329,14 @@ impl Base for ProvingNode {
...
 
     fn handle_connect_failure(&mut self, pub_id: PublicId, outbox: &mut EventBox) -> Transition {
+        debug!("{} received connection failure from {:?}", self, pub_id);
+        let _ = self.resource_proofing_status.remove(&pub_id);
         RelocatedNotEstablished::handle_connect_failure(self, pub_id, outbox)
     }
 
     fn handle_lost_peer(&mut self, pub_id: PublicId, outbox: &mut EventBox) -> Transition {
         debug!("{} Received LostPeer - {}", self, pub_id);
+        let _ = self.resource_proofing_status.remove(&pub_id);
```
Comment (jeanphilippeD) [inline src/states/proving_node.rs]: Why is that need to change?

File: src/states/proving_node.rs [hunk 3]
```diff
@@ -293,6 +377,9 @@ impl Base for ProvingNode {
                     difficulty,
                     log_ident,
                 );
+                if let Some(status) = self.resource_proofing_status.get_mut(&pub_id) {
```
Comment (jeanphilippeD) [inline src/states/proving_node.rs]: Would we be better off using a BTreeSet?

File: tests/mock_crust/mod.rs [hunk 4]
```diff
@@ -172,6 +172,23 @@ fn node_joins_in_front() {
     verify_invariant_for_all_nodes(&mut nodes);
 }
 
+#[test]
+fn joining_node_with_ignoring_candidate_info() {
+    let network = Network::new(MIN_SECTION_SIZE, None);
+    let mut nodes = create_connected_nodes(&network, MIN_SECTION_SIZE);
+    let bootstrap_config = BootstrapConfig::with_contacts(&[nodes[0].handle.endpoint()]);
+    for node in nodes.iter_mut() {
```
Comment (jeanphilippeD) [inline tests/mock_crust/mod.rs]: I think we may only want to do that on some of the nodes.

File: tests/mock_crust/utils.rs [hunk 5]
```diff
@@ -418,6 +418,36 @@ pub fn poll_and_resend_until(
     panic!("poll_and_resend has been called {} times.", MAX_POLL_CALLS);
 }
 
+/// A special advancing pattern to test the ability of resending `CandidateInfo`.
+pub fn poll_and_resend_until_reduced_advance(nodes: &mut [TestNode]) {
```
Comment (jeanphilippeD) [inline tests/mock_crust/utils.rs]: As discussed in mumble, best to not add a duplicate function and make that extra advance clock work in normal poll_and_resend_until.

**Comment by jeanphilippeD** [comment]
Lots of good things, left a few comments. Maybe good to add Jon as reviewer.

File: src/states/proving_node.rs (hunk 2 above)
Comment (maqi) [reply src/states/proving_node.rs]: If it's not received, the disconnected peer will be left over in the `resource_proofing_status` map, which will cause the `connection_info_req` to be sent out forever and exceeding the polling max_limitation

File: src/states/proving_node.rs (hunk 3 above)
Comment (maqi) [reply src/states/proving_node.rs]: we need the status value to decide whether to send out CandidateInfo or not, so I think a map is required?

Comment (jeanphilippeD) [reply src/states/proving_node.rs]: Yes, we are populating the map initially and iterating over it. make sense,=.

File: tests/mock_crust/utils.rs [hunk 6]
```diff
@@ -381,15 +381,17 @@ pub fn poll_all_until(
...
 }
 
 /// Polls and processes all events, until there are no unacknowledged messages left.
 /// should_stop: can be used for an early return from poll_and_resend
+/// extra_advance: this is so far only used for the ignoring candidate_info test.
 pub fn poll_and_resend_until(
     nodes: &mut [TestNode],
     clients: &mut [TestClient],
     should_stop: &Fn(&[TestNode]) -> bool,
+    mut extra_advance: Option<u64>,
```
Comment (jeanphilippeD) [inline tests/mock_crust/utils.rs]: Why do we need that field? which test fail if we do it unconditionally?

Comment (jeanphilippeD) [reply tests/mock_crust/utils.rs]: Lets keep it as you did for now, we can reconsider how best to improve it later.

File: tests/mock_crust/mod.rs [hunk 7]
```diff
@@ -177,15 +180,18 @@ fn joining_node_with_ignoring_candidate_info() {
     let network = Network::new(MIN_SECTION_SIZE, None);
     let mut nodes = create_connected_nodes(&network, MIN_SECTION_SIZE);
     let bootstrap_config = BootstrapConfig::with_contacts(&[nodes[0].handle.endpoint()]);
-    for node in nodes.iter_mut() {
-        node.inner.set_ignore_candidate_info_counter(1);
+    // Make half of the elders ingoring the candidate_info to simulate lagging.
+    // Without the resending, this will block the joining node to be approved.
+    for i in 0..((nodes.len() + 1) / 2) {
```
Comment (jeanphilippeD) [inline tests/mock_crust/mod.rs]: for node in nodes.iter_mut().take(2) {
...
}

File: src/states/elder/mod.rs (hunk 1 above)
Comment (jeanphilippeD) [reply src/states/elder/mod.rs]: I'm confused a bit, so what was the fix for that?

File: src/states/proving_node.rs (hunk 2 above)
Comment (jeanphilippeD) [reply src/states/proving_node.rs]: Why did that change with this PR? why did it not happen before as well?

File: src/states/elder/mod.rs (hunk 1 above)
Comment (maqi) [reply src/states/elder/mod.rs]: This itself is not trying to fix a bug. It's mainly just my over-caustious about the duplicated candidate_info.
I double check the code flow after that point and the case of duplication shall be addressed already.
So I just removed this check here.

File: src/states/proving_node.rs (hunk 2 above)
Comment (maqi) [reply src/states/proving_node.rs]: The `resource_proofing_status` map is introduced by this PR only. 

Previously, a disconnected connection is will have only one `disconnected` msg to be sent to peer.
But now, with the new `resending` attempt, if the entry doesn't get cleaned up, we will send out ConnectionInfo request each time reached timeout.

Comment (jeanphilippeD) [reply src/states/proving_node.rs]: Thanks, yes I got confused there.
still, should it not be set to false instead?

> But now, with the new resending attempt, if the entry doesn't get cleaned up, we will send out ConnectionInfo request each time reached timeout.

Won't it eventually stop when reaching the trying to prove themselves timeout?

Comment (maqi) [reply src/states/proving_node.rs]: >Won't it eventually stop when reaching the trying to prove themselves timeout
No. As long as the node is in the proving_node state and having the map which contains the entry, the connection_info_req will be sent out each time a timeout is reached.

File: tests/mock_crust/utils.rs (hunk 6 above)
Comment (maqi) [reply tests/mock_crust/utils.rs]: >Why do we need that field
that's for the `ignoring_candidate_info` test.
>which test fail if we do it unconditionally
the new `joining_node_with_ignoring_candidate_info` test

File: src/states/proving_node.rs (hunk 2 above)
Comment (jeanphilippeD) [reply src/states/proving_node.rs]: Yes, that is what it is suppose to do, but does it not move to Terminated state?

File: tests/mock_crust/mod.rs (hunk 7 above)
Comment (maqi) [reply tests/mock_crust/mod.rs]: hmm... `2` only works for the `min_network_size` which is 4. 
I thought we want a more general pattern?
how about: `nodes.iter_mut().take((nodes.len() + 1) / 2)` ?

Comment (jeanphilippeD) [reply tests/mock_crust/mod.rs]: Ok with your suggestion, but why the +1?
Why not `take(nodes.len() / 2)`?

File: src/states/proving_node.rs (hunk 2 above)
Comment (maqi) [reply src/states/proving_node.rs]: > does it not move to Terminated state

No. there is no timeout on proving_node itself.
There is only timeout on the elders side which they just reject and disconnect.
But because the proving_node itself keeps sending out msg, the `poll` loop won't break and exceeds the max

File: tests/mock_crust/mod.rs (hunk 7 above)
Comment (maqi) [reply tests/mock_crust/mod.rs]: for `odd` numbers, `5 / 2 = 2`, which is `less than half`
maybe it doesn't matter, as we require supermajority.
I will use `nodes.len() / 2` then.

File: tests/mock_crust/mod.rs [hunk 8]
```diff
@@ -172,6 +172,26 @@ fn node_joins_in_front() {
...
+    for node in nodes.iter_mut().take(MIN_SECTION_SIZE / 2) {
+        node.inner.set_ignore_candidate_info_counter(1);
+    }
+    nodes.push(
+        TestNode::builder(&network)
+            .bootstrap_config(bootstrap_config.clone())
+            .create(),
+    );
+    let dummy = |_nodes: &[TestNode]| false;
+    poll_and_resend_until(&mut nodes, &mut [], &dummy);
```
Comment (jeanphilippeD) [inline tests/mock_crust/mod.rs]: Can use poll_and_resend now :-)

File: src/states/elder/mod.rs (hunk 1 above)
Comment (jeanphilippeD) [reply src/states/elder/mod.rs]: I forgot and meant to say:
This is a good place for a unit tests. You add a new test that follow the pattern of the existing one to exercise this corner case and verify it does the right thing.
